src/v/net/connection_rate.cc

```diff
@@ -0,0 +1,236 @@
+    update_tokens_if_needed(sem);
+    co_await sem->rate.wait(_max_wait_time, 1);
+    spawn_updating_fiber_if_needed(sem);
+}
+
+connection_rate::connection_rate_t
+connection_rate::find_sem(const ss::sstring& ip) {
+    auto ovverides_it = _overrides.find(ip);
+    if (ovverides_it != _overrides.end()) {
+        return ovverides_it->second;
+    }
+
+    return _general_rate;
+}
+
+void connection_rate::stop() {
+    if (_general_rate) {
+        _general_rate->broke_semaphore();
+    }
+    for (auto& [_, sem] : _overrides) {
+        sem->broke_semaphore();
+    }
+}
+
+void connection_rate::update_tokens_if_needed(connection_rate_t rate_counter) {
+    int64_t max_tokens_for_update = rate_counter->max_tokens
+                                    - rate_counter->rate.current();
+
+    if (max_tokens_for_update == 0) {
+        return;
+    }
+
+    auto now = ss::lowres_clock::now();
+    auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(
+      now - rate_counter->last_update_time);
+
+    int64_t tokens_for_update = duration.count()
+                                / rate_counter->one_token_time.count();
+
+    if (tokens_for_update == 0) {
+        return;
+    }
+
+    rate_counter->last_update_time = now;
+
+    int64_t new_tokens_count = std::min(
+      tokens_for_update, max_tokens_for_update);
+
+    rate_counter->rate.signal(new_tokens_count);
+}
+
+void connection_rate::spawn_updating_fiber_if_needed(
+  connection_rate_t rate_counter) {
+    if (rate_counter->rate.current() == 0) {
+        // Should spawn fiber to update tokens on next second
+        ssx::spawn_with_gate(
+          _connection_gate,
+          [this,
+           rate_counter,
+           sleep_time = rate_counter->one_token_time]() mutable {
+              return ss::sleep(ss::lowres_clock::duration(sleep_time))
+                .then([this, rate_counter] {
+                    if (rate_counter.get() != nullptr) {
+                        update_tokens_force(rate_counter);
+                    }
+                });
+          });
+    }
+}
+
+void connection_rate::update_tokens_force(connection_rate_t rate_counter) {
+    auto now = ss::lowres_clock::now();
+    auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(
+      now - rate_counter->last_update_time);
+
+    // If somebody update tokens in updating interval we should skip
+    // updating phase
+    if (
+      duration.count() < rate_counter->one_token_time.count()
+      || rate_counter->rate.current() > 0) {
+        return;
+    }
+
+    rate_counter->last_update_time = now;
+    rate_counter->rate.signal(1);
+}
+
+void connection_rate::update_connection_rate(
+  connection_rate_t rate_counter, std::optional<int64_t> new_value) {
+    if (new_value) {
+        auto should_spawn_fiber = rate_counter->update_rate(new_value.value());
+        if (should_spawn_fiber == should_spawn_fiber_on_update::yes) {
+            spawn_updating_fiber_if_needed(rate_counter);
```

Comment:
Also we can signal(1) after updating to allow one connection comes in and spawn fiber if needed